### PR TITLE
feat(tex): improve texture optimization

### PR DESCRIPTION
**Pull request by mklokocka**
Introduced some further improvements to the texture optimization process:

1) Do not decompress images when there is no need, that is, when resizing is required.
There is no need for decompressing and compressing an image again if resizing is not applied.
2) Enforce conversion of results when input has no/opaque alpha to a format without alpha (either compressed, or not).
This fixes the issue with `Resize` introducing full white alpha to the image, which can break stuff like normal maps. Normal maps with full white alpha are not the same as those with no alpha (alpha represents specularity).
3) Never remove mipmaps - if resizing is applied and the input had mipmaps, regenerate them after resize.
This is because mipmaps are very important to game visuals and performance. Removing them while resizing doesn't make sense even if mipmaps are not requested.

An important point: recompressing with a "higher" BC does not improve quality. If the input is already block compressed, decompressing does not restore quality, and thus compressing it again with better BC does not improve on the input at all. In fact, it only increased the size of the resulting image without any benefits.

### CODE REVIEW

File: src/tex/optimize.cpp [hunk 1]
```diff
@@ -157,14 +158,18 @@ auto compute_optimization_steps(const Texture &file, const Settings &sets) noexc
...
         res.mipmaps = true;
 
     // I prefer to keep steps independent, but this one has to depend on add_transparent_alpha. If we add an alpha, the output format must have alpha
     res.best_format = best_output_format(file,
                                          sets,
                                          res.add_transparent_alpha);
 
+    // Ensure result is saved without alpha in cases where it could be problematic (i.e. resizing).
+    if (res.resize && res.best_format == sets.output_format.uncompressed_without_alpha)
+        res.convert = true;
```
Comment (Guekka) [inline src/tex/optimize.cpp]: You are only checking uncompressed. I think it also applies to compressed?

**Review by Guekka — COMMENTED** [review]
I need more time to review the new tests, but it looks good

File: src/tex/optimize.cpp (hunk 1 above)
Comment (mklokocka) [reply src/tex/optimize.cpp]: We should convert/compress on every resize, yes, because for compressed textures resizing requires decompressing. But this is implicit in the `optimize` function already (`must_decompress` flag), even if `res.convert == false`.

This is why I am thinking about refactoring the whole steps -> optimize logic a bit. They kinda clash into each other with their responsibilites. I'm just not sure how I would do it (for example the decompression requirement - should it be part of the optimization steps or the optimize function? I think both are justifiable).

Comment (Guekka) [reply src/tex/optimize.cpp]: Please add a comment explaining this

The reason for `compute_optimization_steps` to be separated is to allow the caller to change the steps and to allow dry run. CAO5 logic was a mess, so this is what I came up with. However, I agree that it will get hard to follow the more steps we add

Comment (mklokocka) [reply src/tex/optimize.cpp]: Yeah, it's good to have it separated especially if a fine-grain per-file control will be added one day (as we discussed on Discord).

I actually went back and rechecked everything and decided that this conversion should happen implicitely. Also, I'm dumb and didn't know all transformations including mipmap generation requires decompressed texture, so yeah, fixed that.

Added a few more tests to be sure this time.

File: src/tex/optimize.cpp [hunk 2]
```diff
@@ -16,10 +16,14 @@ namespace btu::tex {
...
-    auto res              = Result{std::move(file)};
-
-    if (compressed)
+    const auto &info           = file.get().GetMetadata();
+    const auto must_decompress = DirectX::IsCompressed(info.format);
+    // Special case - force conversion if result shouldn't have alpha to get rid of alpha bits that are added by DirectX.
+    const auto should_convert  = sets.convert || must_decompress || !DirectX::HasAlpha(sets.best_format);
+    auto res                   = Result{std::move(file)};
+
+    // We must only decompress if we want to resize or transform the image.
```
Comment (Guekka) [inline src/tex/optimize.cpp]: This comment is wrong

**Review by Guekka — CHANGES_REQUESTED** [review]

File: src/tex/optimize.cpp (hunk 2 above)
Comment (mklokocka) [reply src/tex/optimize.cpp]: Changed and moved to the variable definition.

**Comment by mklokocka** [comment]
Not sure why the pipeline failed this time, nothing changed besides a comment.

**Comment by Guekka** [comment]
> Not sure why the pipeline failed this time, nothing changed besides a comment.

Looks like a spurious fail

**Review by Guekka — APPROVED** [review]
